File: DSA/C++/Postorder Traversal on Binary Tree (Non recursive).cpp
// In this program we will perform postorder traversal on binary tree (non recursively) or we can say using stack

#include <iostream>
using namespace std;

class Node {
public:
  int val;
  Node* left;
  Node* right;
    Node (int data){
      val = data;
      left = NULL;
      right = NULL;
    }
};

Node* BuildTree(Node* &root){
  int data;
  cout<<"Enter data:- ";
  cin>>data;

  if (data==-1) return NULL;

  root = new Node(data);

  cout<<"Enter data in left of "<<data<<endl;
  root->left = BuildTree(root->left);

  cout<<"Enter data in right of "<<data<<endl;
  root->right = BuildTree(root->right);

  return root;

}

class Stack{
public:
  Node* st[100];
  int top = -1;

  void push(Node* &root){
    st[++top] = root;
  }

  Node* pop(){
    return st[top--];
  }
};

void Postorder(Node* root) {
    if (root == NULL) return;

    Node* current = root;
    Node* lastVisited = NULL;
    Node* stack[100]; 
    int top = -1;     

    while (current != NULL || top != -1) {
        while (current != NULL) {
            stack[++top] = current; 
            current = current->left; 
        }

        current = stack[top];

        if (current->right != NULL && current->right != lastVisited) {
            current = current->right; 
        } else {
            cout << current->val << " ";
            lastVisited = current; 
            current = NULL; 
            top--; 
        }
    }
}


int main() {
  Node* root = NULL;
  BuildTree(root);
  Postorder(root);
}

File: DSA/Python/Postorder Traversal on Binary Tree (Non recursive).py
# In this program we will perform postorder traversal on a binary tree (non-recursively) using a stack

class Node:
    def __init__(self, data):
        self.val = data
        self.left = None
        self.right = None

def build_tree():
    data = int(input("Enter data: "))
    
    if data == -1:
        return None
    
    root = Node(data)
    
    print(f"Enter data in left of {data}")
    root.left = build_tree()
    
    print(f"Enter data in right of {data}")
    root.right = build_tree()
    
    return root

def postorder_non_recursive(root):
    if root is None:
        return
    
    stack = []
    last_visited = None
    current = root
    
    while stack or current:
        while current:
            stack.append(current)
            current = current.left
        
        peek_node = stack[-1]
        
        if peek_node.right and last_visited != peek_node.right:
            current = peek_node.right
        else:
            print(peek_node.val, end=" ")
            last_visited = stack.pop()

root = build_tree()
print("Postorder Traversal (Non-Recursive):", end=" ")
postorder_non_recursive(root)
print()
